timesketch/lib/datastores/opensearch.py

```diff
@@ -1271,3 +1296,505 @@ def flush_queued_events(self, retry_count=0):
+        view of the data across multiple paginated and sliced requests.
+
+        This method is designed for efficient bulk export of large datasets.
+        It yields individual event dictionaries, including '_id' and '_index'.
+
+        The following parameters can be overridden by the caller. If not provided,
+        they will default to values from the Timesketch configuration or
+        sensible internal defaults:
+            - `page_size`: Defaults to `OPENSEARCH_SLICED_EXPORT_DEFAULT_PAGE_SIZE`
+            - `pit_keep_alive`: Defaults to `OPENSEARCH_SLICED_EXPORT_PIT_KEEP_ALIVE`
+            - `num_slices`: Defaults to `OPENSEARCH_SLICED_EXPORT_NUM_SLICES` or 4
+            - `request_timeout_per_slice`: Defaults to
+              `OPENSEARCH_SLICED_EXPORT_REQUEST_TIMEOUT` or a calculated value
+              based on the general OpenSearch timeout +20s.
+
+        The internal queue size is determined by `page_size`, `num_slices`, and
+        the `OPENSEARCH_SLICED_EXPORT_QUEUE_BUFFER_FACTOR` configuration.
+
+        Args:
+            indices_for_pit: List of index names to query.
+            base_query_body: The base OpenSearch query body (must not include
+                             size, sort, pit, or slice clauses). Can include
+                             "_source" to specify fields to return.
+            sort_criteria: Optional. The sort criteria for search_after.
+                           If None, defaults to `[{"_id": "asc"}]`.
+                           Must include a unique tie-breaker field (e.g. _id)
+                           for stable results across pages.
+            page_size: Optional. Number of documents per request per slice.
+                       Overrides the global default (10_000) if provided.
+            pit_keep_alive: Optional. Keep-alive duration for the Point-In-Time context
+                            (e.g., "5m"). Overrides the global default if provided.
+            num_slices: Optional. Number of parallel slices to use for the export.
+                        Overrides the global default if provided.
+                        Must be between 1 and 1024.
+            request_timeout_per_slice: Optional timeout in seconds for individual
+                                       OpenSearch search requests within each slice.
+                                       Overrides the global default (30s) if provided.
+
+        Yields:
+            Dict: Individual event documents (including _id and _index from the hit).
+
+        Raises:
+            ValueError: If indices_for_pit is empty or num_slices is invalid.
+            errors.DataStoreQueryError: For critical issues during the export.
+        """
+        if not indices_for_pit:
+            raise ValueError("indices_for_pit cannot be empty.")
+
+        # Getting default values if not provided otherwise by the caller.
+        effective_sort_criteria = sort_criteria or _DEFAULT_PIT_SORT_CRITERIA
+        effective_page_size = (
```

Comment:
also, can page:size be negative? if not, consider checking

Reply:
What the page_size can and cannot be is basically handled by the opensearch API. If you set it to negative (or larger than your opensearch instance is configured to handle) the opensearch client will raise an error with the search and halt all exports.

Reply:
ACK